### PR TITLE
Back-porting of Zombie SNAP improvements

### PR DESCRIPTION
**Pull request by athomps**
**Summary**

Completed back-porting of Zombie SNAP improvements, particularly noteworthy is reduction in memory footprint, elimination of most multidimensional arrays, elimination of diagonal_style, elimination of Z array in force calculation, gives about 2x speed-up on typical benchmarks.

**Related Issues**

**Author(s)**

Aidan Thompson

**Licensing**

**Backward Compatibility**

SNAP param files that contain the diagonalstyle keyword will generate this error:

ERROR: Incorrect SNAP parameter file

Solution is to remove the diagonalstyle keyword.  The code now supports only diagonalstyle=3.  This change has been documented.

**Implementation Notes**

The three SNAP potentials in examples/snap work as before.  Additional testing of old versus new on output from compute sna, snad, snav, and quadratic SNAP was performed, with only tiny differences < 1e-10.

**Post Submission Checklist**

_Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply_

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] A package specific README file has been included or updated
- [x] One or more example input decks are included

**Further Information, Files, and Links**

_Put any additional information here, attach relevant text or image files, and URLs to external sites (e.g. DOIs or webpages)_

### CODE REVIEW

**Comment by akohlmey** [comment]
@athomps this PR seems to break the KOKKOS version of SNAP.  There also seems to be a typo in the docs ("ompute" instead of "compute", if i read the error logs correctly).

**Comment by akohlmey** [comment]
@athomps BTW, when editing the checklist, you must delete the extra blank so it renders the markup correctly. or better yet, set the entries to enabled by clicking on the buttons **after** submitting the PR.

**Comment by athomps** [comment]
I updated KOKKOS package. I did not find the ompute typo.

File: doc/src/compute_sna_atom.txt [hunk 1]
```diff
@@ -192,25 +187,20 @@ command that includes all pairs in the neighbor list.
...
+    for j2 in range(0,j1+1):
+        for j in range(j1-j2,min(twojmax,j1+j2)+1,2):
+            if (j>=j1): print j1/2.,j2/2.,j/2. :pre
+
+NOTE: the {diagonal} keyword allowing other possible choices
+for the number of bispectrum components was removed in 2019, 
+since all potentials use the value of 3, corresponding to the
+above set of bispectrum components.
+
+ompute {snad/atom} evaluates a per-atom array. The columns are
```
Comment (akohlmey) [inline doc/src/compute_sna_atom.txt]: @athomps there is the typo

File: src/SNAP/pair_snap.cpp [hunk 2]
```diff
@@ -233,66 +162,22 @@ void PairSNAP::compute_regular(int eflag, int vflag)
       }
     }
 
-    // compute Ui, Zi, and Bi for atom I
+    // compute Ui, Yi for atom I
 
     snaptr->compute_ui(ninside);
```
Comment (stanmoore1) [inline src/SNAP/pair_snap.cpp]: @athomps Isn't this redundant if `quadratic_flag` is set? In other words it should be conditional on `!quadratic_flag`?

**Comment by athomps** [comment]
I don't think it is redundant in the current set up, because U is required to get Y which is required to get deidrj, even with quadratic, but maybe there is a way to exploit it.  I attach a quadratic test script that you can experiment with.

Aidan


--
      Aidan P. Thompson
      01444 Multiscale Science
      Sandia National Laboratories
      PO Box 5800, MS 1322      Phone: 505-844-9702
      Albuquerque, NM 87185     Fax  : 505-845-7442
      E-mail:athomps@sandia.gov Cell : 505-218-1011

**Comment by stanmoore1** [comment]
I mean for `quadratic`, `compute_ui` is already called previously in `compute_bispectrum()`, so it is getting called twice.

**Comment by athomps** [comment]
Yes, but compute_bispectrum is for all atoms, not just one. So we would have to store Ui for all atoms. If that speeds up qaudratic, it is worth doing, as we are starting to use quadratic for real potentials.

Aidan


--
      Aidan P. Thompson
      01444 Multiscale Science
      Sandia National Laboratories
      PO Box 5800, MS 1322      Phone: 505-844-9702
      Albuquerque, NM 87185     Fax  : 505-845-7442
      E-mail:athomps@sandia.gov Cell : 505-218-1011

**Comment by stanmoore1** [comment]
That makes sense, thanks

**Comment by athomps** [comment]
I did a test of making compute_bispectrum conditional on eflag, not quadraticflag, which gives somewhat incorrect results, but is representative. The time reduction was about 20%.  Probably not worth exploiting by itself, but something to keep in mind.

**Comment by stanmoore1** [comment]
@athomps lines 613-614 of `sna.cpp` look like no-ops. Can you confirm? See https://github.com/lammps/lammps/blob/master/src/SNAP/sna.cpp#L613

**Comment by athomps** [comment]
Yes, I remember weighing whether to add or delete those lines.  They have no effect or function, except to satisfy some deep psychological need on my part.  Feel free to delete.

Sent from my iPhone

On Jun 21, 2019, at 10:21 AM, Stan Moore <notifications@github.com<mailto:notifications@github.com>> wrote:


@athomps<https://github.com/athomps> lines 613-614 of sna.cpp look like no-ops. Can you confirm? See https://github.com/lammps/lammps/blob/master/src/SNAP/sna.cpp#L613

**Pull request merged**